src/work/BasicWork.cpp

```diff
@@ -0,0 +1,351 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "work/BasicWork.h"
+#include "lib/util/format.h"
+#include "util/Logging.h"
+#include "util/Math.h"
+
+namespace stellar
+{
+
+size_t const BasicWork::RETRY_NEVER = 0;
+size_t const BasicWork::RETRY_ONCE = 1;
+size_t const BasicWork::RETRY_A_FEW = 5;
+size_t const BasicWork::RETRY_A_LOT = 32;
+size_t const BasicWork::RETRY_FOREVER = 0xffffffff;
+
+std::set<BasicWork::Transition> const BasicWork::ALLOWED_TRANSITIONS = {
+    Transition(InternalState::PENDING, InternalState::RUNNING),
+    Transition(InternalState::PENDING, InternalState::DESTRUCTING),
+    Transition(InternalState::RUNNING, InternalState::RUNNING),
+    Transition(InternalState::RUNNING, InternalState::WAITING),
+    Transition(InternalState::RUNNING, InternalState::SUCCESS),
+    Transition(InternalState::RUNNING, InternalState::FAILURE),
+    Transition(InternalState::RUNNING, InternalState::RETRYING),
+    Transition(InternalState::RUNNING, InternalState::DESTRUCTING),
+    Transition(InternalState::WAITING, InternalState::RUNNING),
+    Transition(InternalState::WAITING, InternalState::DESTRUCTING),
+    Transition(InternalState::RETRYING, InternalState::WAITING),
+    Transition(InternalState::SUCCESS, InternalState::DESTRUCTING),
+    Transition(InternalState::SUCCESS, InternalState::PENDING),
+    Transition(InternalState::FAILURE, InternalState::DESTRUCTING),
+    Transition(InternalState::FAILURE, InternalState::PENDING),
+};
+
+BasicWork::BasicWork(Application& app, std::string name, size_t maxRetries)
+    : mApp(app), mName(std::move(name)), mMaxRetries(maxRetries)
+{
+}
+
+BasicWork::~BasicWork()
+{
+}
+
+void
+BasicWork::shutdown()
+{
+    mState = InternalState::DESTRUCTING;
```

Comment:
Why is setState not used here? I think the reason should be added as a comment here.

Reply:
It probably should be used here ;)